cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -105,13 +113,113 @@ func ValidateAuthorizationMode(authzMode string, fldPath *field.Path) field.Erro
 	return field.ErrorList{}
 }
 
+func ValidateDiscovery(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidateTLSBootstrapToken(c, fldPath)...)
+	var count int
+	if len(c.DiscoveryToken) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryToken(c, fldPath)...)
+		count++
+	}
+	if len(c.DiscoveryFile) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryFile(c, fldPath)...)
+		count++
+	}
+	if len(c.DiscoveryURL) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryURL(c, fldPath)...)
+		count++
+	}
+	if count > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "exactly one discovery strategy can be provided"))
+	}
+	return allErrs
+}
+
+func ValidateTLSBootstrapToken(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(cfg.Token) != 0 && (len(cfg.TLSBootstrapToken) != 0 || len(cfg.DiscoveryToken) != 0) {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "--token is mutually exclusive with --bootstrap-token and --discovery-token"))
+	}
+	if len(cfg.Token) == 0 && (len(cfg.TLSBootstrapToken) == 0 || len(cfg.DiscoveryToken) == 0) {
+		allErrs = append(
+			allErrs,
+			field.Invalid(fldPath, nil, "If --token is not specified, both --bootstrap-token and --discovery-token have to be provided"),
+		)
+	}
+	if len(cfg.Token) != 0 {
```

Comment:
I think that `--tls-bootstrap-token` should only be available with `--discovery-token`.
With the file or URL method, it's expected that the auth information already is in the file (=bootstrap-token for example)